lib/rules/no-unused-vars.js

```diff
@@ -858,10 +861,439 @@ module.exports = {
+
+                                                // var {a} = foo;
+                                                if (parent.parent.parent.type === "VariableDeclarator") {
+                                                    if (parent.parent.parent.parent.declarations.length === 1) {
+                                                        return fixer.removeRange(parent.parent.parent.parent.range);
+                                                    }
+
+                                                    if (getTokenBeforeValue(parent.parent.parent) === ",") {
+                                                        return fixer.removeRange(
+                                                            [getBeforeToken(parent.parent.parent), parent.parent.parent.range[1]]
+                                                        );
+                                                    }
+
+                                                    return fixer.removeRange(
+                                                        [parent.parent.parent.range[0], getAfterToken(parent.parent.parent)]
+                                                    );
+                                                }
+
+                                                // function a({a}) {}
+                                                if (isTypeFunction(parent.parent.parent)) {
+                                                    if (parent.parent.parent.params.length === 1) {
+                                                        return fixer.removeRange(parent.parent.range);
+                                                    }
+
+                                                    if (
+                                                        getTokenBeforeValue(parent.parent) === "(" &&
+                                                        getTokenAfterValue(parent.parent) === ","
+                                                    ) {
+                                                        return fixer.removeRange([parent.parent.range[0], getAfterToken(parent.parent)]);
+                                                    }
+
+                                                    return fixer.removeRange([getBeforeToken(parent.parent), parent.parent.range[1]]);
+                                                }
+
+                                                // fix const { a: { b } } = foo;
+                                                if (getTokenBeforeValue(parent.parent) === ":") {
+                                                    if (parent.parent.parent.parent.type === "ObjectPattern") {
+                                                        if (parent.parent.parent.parent.properties.length === 1) {
+                                                            return fixVariables(parent.parent.parent.parent);
+                                                        }
```

Comment:
There seems to be a lot of repeated code for different combinations of patterns in patterns. Can we maybe unify the code? For example, when we determine that something can be removed, recursively check the parent and so on until we find the uppermost node that can be removed.

Reply:
is this statement suggesting to use loops? may be i don't get it properly.
how about if we use function to reduce the repetition.

Reply:
Yes, loops or recursion. That would avoid code repetition and cover more cases.

For example, the current implementation provides a suggestion to fix `const [[foo]] = bar;` because it has a branch that specifically covers the array pattern in array pattern case. But it doesn't provide a suggestion to fix `const [[[foo]]] = bar;` although it's basically the same case, just one more level deep. The same for any combination of patterns, for example the current implementation provides a suggestion to fix `const { a: [b] } = c;`, but doesn't provide a suggestion to fix `const [{ a: [b] }] = c;`.

Reply:
fix for some nested patterns has implemented, working on some others, so marked it as draft.